internal/querycoord/index_checker.go

```diff
@@ -232,40 +274,34 @@ func (ic *IndexChecker) processHandoffAfterIndexDone() {
 		select {
 		case <-ic.ctx.Done():
 			return
-		case segmentInfo := <-ic.indexedSegmentsChan:
+		case task := <-ic.handoffTaskChan:
+			segmentInfo := task.HandoffSegmentsRequest.SegmentInfos[0]
 			collectionID := segmentInfo.CollectionID
 			partitionID := segmentInfo.PartitionID
 			segmentID := segmentInfo.SegmentID
+
 			log.Info("processHandoffAfterIndexDone: handoff segment start", zap.Any("segmentInfo", segmentInfo))
-			baseTask := newBaseTask(ic.ctx, querypb.TriggerCondition_Handoff)
-			handoffReq := &querypb.HandoffSegmentsRequest{
-				Base: &commonpb.MsgBase{
-					MsgType: commonpb.MsgType_HandoffSegments,
-				},
-				SegmentInfos: []*querypb.SegmentInfo{segmentInfo},
-			}
-			handoffTask := &handoffTask{
-				baseTask:               baseTask,
-				HandoffSegmentsRequest: handoffReq,
-				broker:                 ic.broker,
-				cluster:                ic.cluster,
-				meta:                   ic.meta,
-			}
-			err := ic.scheduler.Enqueue(handoffTask)
+			err := ic.scheduler.Enqueue(task)
 			if err != nil {
 				log.Error("processHandoffAfterIndexDone: handoffTask enqueue failed", zap.Error(err))
 				panic(err)
 			}
 
-			go func() {
-				err := handoffTask.waitToFinish()
-				if err != nil {
+			go func(task *handoffTask) {
+				// release lock on segment
+				segmentID := task.HandoffSegmentsRequest.SegmentInfos[0].SegmentID
+				if err := ic.ReleaseSegmentLock(ic.ctx, task.taskID, []UniqueID{segmentID}); err != nil {
+					log.Warn("release segment reference lock failed", zap.Int64("task id", task.taskID), zap.Int64("segmentID", segmentID))
+					panic(err)
+				}
+
+				if err := task.waitToFinish(); err != nil {
```

Comment:
Why release the lock first, then wait for the task to finish? If compaction occurred after releasing lock and before the task finishing, can you make ensure the task is done?